packages/task/src/browser/task-service.ts

```diff
@@ -998,13 +1003,16 @@ export class TaskService implements TaskConfigurationClient {
              *       Reason: Maybe a new task type wants to also be displayed in a terminal.
              */
             if (typeof taskInfo.terminalId === 'number') {
-                this.attach(taskInfo.terminalId, taskInfo.taskId);
+                await this.attach(taskInfo.terminalId, taskInfo);
             }
             return taskInfo;
         } catch (error) {
             const errorStr = `Error launching task '${taskLabel}': ${error.message}`;
             this.logger.error(errorStr);
             this.messageService.error(errorStr);
+            if (taskInfo && typeof taskInfo.terminalId === 'number') {
+                this.shellTerminalServer.onFailedAttach(taskInfo.terminalId);
+            }
```

Comment:
> Is there any reason to worry that someone might start terminal processes that never get attached to? This seems bind the behavior of the `TerminalProcess` fairly tightly to a set of expectations that may only apply to the task system.

Added this part to unregister on attach failed, This is the only place where `this.taskServer.run` is executed so hopefully there won't be processes that  will stay always registered...

Reply:
> Added this part to unregister on attach failed, This is the only place where `this.taskServer.run` is executed so hopefully there won't be processes that will stay always registered...

I think this is definitely good, but I still have some misgivings about this system. Although the `TerminalProcess` class is only used in Theia for tasks, so that accounting for cleanup in the context of task terminal interactions is sufficient upstream, it's possible that downstream adopters could have used `TerminalProcess`es for other purposes and relied on the fact that those processes would automatically clean themselves up on exit. The existing code on `master` isn't very well designed for extension - too much work is done in the constructor instead of delegating to named methods - but might it be possible to create a new class derived from `TerminalProcess` that has the custom logic that waits for attachment before unregistering while leaving the functionality of the ancestors intact?

Reply:
@colin-grant-work Thanks for the idea, Updated the code, Can you review and check that it's okay now?